### ISSUE TITLE
Impersonation for Databricks

### ISSUE THREAD

**Issue by paoliniluis** [metabase/metabase#48438]
**Is your feature request related to a problem? Please describe.**
A user mentioned that they define everything on the Databricks unity catalog, so they want Metabase to inherit those permissions for the queries

**Describe the solution you'd like**
Use https://www.databricks.com/product/unity-catalog for permissions

**Describe alternatives you've considered**
NA

**How important is this feature to you?**
Requested by a lead (banking)

**Additional context**
NA

**Comment by rileythomp** [comment]
We'll use DB routing to achieve impersonation for Databricks. You can add a routing DB with the same host/http path/catalogs/schemas with the relevant credentials.

**Comment by Gerrit-K** [comment]
Hey @rileythomp we're eagerly waiting on this feature and are happy to hear that there's progress! Could you perhaps elaborate a bit on your comment? Does it mean that it's already possible to set up impersonation using [database routing](https://www.metabase.com/docs/latest/permissions/database-routing) or will you just use the same mechanism behind the scenes? And if there's still implementation work to be done, is there another issue we can follow to track progress? Thanks!

**Comment by rileythomp** [comment]
@Gerrit-K 

> Does it mean that it's already possible to set up impersonation using [database routing](https://www.metabase.com/docs/latest/permissions/database-routing)

It's more that you can get the equivalent of impersonation for Databricks through DB routing. Databricks doesn't support RBAC via `SET ROLE` statements which is how impersonation works for other drivers. It uses identities (e.g. users, service principals) with their own credentials. So you can get the equivalent of impersonation by using DB routing with the same DB connection details and the relevant credentials. 

Normal impersonation flow: 
1. Set a `(database, group)`'s `View Data` permissions to `Impersonated` with a user attribute
2. Give users in that group the user attribute with the value corresponding to a DB role
3. Queries are executed with the role set for the user

Databricks "impersonation"/DB routing flow:
1. Enable DB routing and select a user attribute for it
2. Add routing DBs with the same DB connection details (host, http path, etc.), the auth credentials for a user/service principal, and a slug (i.e. a user/service principal name)
3. Give users the DB routing user attribute with the value corresponding to a routing DB slug
4. Queries are executed with the routing credentials for the user

Let me know if that makes sense

**Comment by Gerrit-K** [comment]
@rileythomp Thanks for the explanation. It does make sense, but I'm afraid it wouldn't be an equivalent of impersonation for us.

> 2. Add routing DBs with the same DB connection details (host, http path, etc.), the auth credentials for a user/service principal, and a slug (i.e. a user/service principal name)

If I understand this correctly, this would require us to create a connection (+ PAT or SP credentials) _for every user_ or at least every distinct set of possible roles a user can have (whichever has the lower cardinality). 

I could imagine this works just fine if you have a small set of roles and only a single role per user, e.g.
- User A role: admin -> access to everything
- User B role: hr -> access to hr tables
- User C role: sales -> access to sales tables
- etc.

But our use-case looks a bit different. We use roles to directly define access to data assets, e.g.
- User A roles: hr-admin,hr-viewer,sales-admin,sales-viewer,... 
- User B roles: hr-viewer
- User C role: sales-viewer,sales-admin

Of course this is just a simplified example - imagine having tens to hundreds of roles and arbitrary combinations of them for every user depending on their job function. This just wouldn't work with what I understand from the database routing. With the "real" connection impersonation, however, it's as simple as `SET ROLE hr-admin,hr-viewer,sales-admin,sales-viewer,...`.

As far as I'm aware, other tools like Tableau, Power BI or Looker solve this via U2M OAuth / OBO tokens. I.e. they redirect the user to Databricks to authorize Metabase to perform queries on their behalf. Metabase then gets a token for that specific user which it can use for all subsequent queries. I think this would be almost identical to the database routing, except that it doesn't have the maintenance overhead of creating all the credentials and connections, which would scale much better. Have you considered this option yet?

**Comment by dragonsahead** [comment]
@Gerrit-K we need to develop OAuth for authentication into databases first, and then add a feature to create a connection pool PER TOKEN (kinda like a route per user). It’s a major project

**Comment by Gerrit-K** [comment]
@dragonsahead Thanks! I do understand that this is no simple feat. Yet, I don't feel like "connection impersonation" is currently fully addressed with this. Since the ticket is closed, I would like to know if this is a "won't do" or a "will take some time". In case of the former, we have a serious feature gap that will not be addressed.

**Comment by dragonsahead** [comment]
@Gerrit-K the issue is similar/same as https://github.com/metabase/metabase/issues/34932 and https://github.com/metabase/metabase/issues/13181, but this time for Databricks (we might have to create a new issue to map your request to that). The feature will take some time tho

**Comment by rileythomp** [comment]
@Gerrit-K I've created this issue for Databricks OAuth: https://github.com/metabase/metabase/issues/69349. Not sure when we will get to it. 
@dragonsahead Feel free to add more detail to it
